zcash_primitives/src/transaction/mod.rs

```diff
@@ -53,6 +53,91 @@ impl fmt::Display for TxId {
     }
 }
 
+/// The set of defined transaction format versions.
+///
+/// This is serialized in the first four or eight bytes of the transaction format, and
+/// represents valid combinations of the `(overwintered, version, version_group_id)`
+/// transaction fields.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum TxVersion {
+    Sprout(u32),
+    Overwinter,
+    Sapling,
+    #[cfg(feature = "zfuture")]
+    ZFuture,
+}
+
+impl TxVersion {
+    pub fn read<R: Read>(mut reader: R) -> io::Result<Self> {
+        let header = reader.read_u32::<LittleEndian>()?;
+        let overwintered = (header >> 31) == 1;
+        let version = header & 0x7FFFFFFF;
+
+        if overwintered {
+            match (version, reader.read_u32::<LittleEndian>()?) {
+                (OVERWINTER_TX_VERSION, OVERWINTER_VERSION_GROUP_ID) => Ok(TxVersion::Overwinter),
+                (SAPLING_TX_VERSION, SAPLING_VERSION_GROUP_ID) => Ok(TxVersion::Sapling),
+                #[cfg(feature = "zfuture")]
+                (ZFUTURE_TX_VERSION, ZFUTURE_VERSION_GROUP_ID) => Ok(TxVersion::ZFuture),
+                _ => Err(io::Error::new(
+                    io::ErrorKind::InvalidInput,
+                    "Unknown transaction format",
+                )),
+            }
+        } else {
+            Ok(TxVersion::Sprout(version))
+        }
+    }
+
+    pub fn header(&self) -> u32 {
+        // After Sprout, the overwintered bit is always set.
+        let overwintered = match self {
+            TxVersion::Sprout(_) => 0,
+            _ => 1 << 31,
+        };
+
+        overwintered | match self {
+            TxVersion::Sprout(v) => *v,
+            TxVersion::Overwinter => OVERWINTER_TX_VERSION,
+            TxVersion::Sapling => SAPLING_TX_VERSION,
+            #[cfg(feature = "zfuture")]
+            TxVersion::ZFuture => ZFUTURE_TX_VERSION,
+        } 
+    }
+
+    pub fn version_group_id(&self) -> u32 {
+        match self {
+            TxVersion::Sprout(_) => 0,
+            TxVersion::Overwinter => OVERWINTER_VERSION_GROUP_ID,
+            TxVersion::Sapling => SAPLING_VERSION_GROUP_ID,
+            #[cfg(feature = "zfuture")]
+            TxVersion::ZFuture => ZFUTURE_VERSION_GROUP_ID,
+        }
+    }
+
+    pub fn write<W: Write>(&self, mut writer: W) -> io::Result<()> {
+        writer.write_u32::<LittleEndian>(self.header())?;
+        match self {
+            TxVersion::Sprout(_) => Ok(()),
+            _ => writer.write_u32::<LittleEndian>(self.version_group_id()),
+        }
+    }
+
+    pub fn has_sprout(&self) -> bool {
+        match self {
+            TxVersion::Sprout(v) => *v >= 2u32,
+            _ => true,
+        }
+    }
+
+    pub fn uses_groth_proofs(&self) -> bool {
+        match self {
+            TxVersion::Sprout(_) | TxVersion::Overwinter => false,
+            _ => true,
```

Comment:
nit: Rather than `_` we should explicitly put

```rust
    Overwinter | Sapling => true,
    #[cfg(feature = "zfuture")]
    ZFuture => true,
```

because if we add a new `TxVersion` variant we'll want to be reminded by the compiler to revisit the correctness of this logic for that variant.

Reply:
Good suggestion. I've made the same change for `has_sprout` as well - for that eventual day when it can once again return `false` :)